cmd/plugins/balloons/policy/balloons-policy.go

```diff
@@ -1417,11 +1428,93 @@ func (p *balloons) pinCpuMem(c cache.Container, cpus cpuset.CPUSet, mems idset.I
 	if p.bpoptions.PinMemory == nil || *p.bpoptions.PinMemory {
 		if c.PreserveMemoryResources() {
 			log.Debug("  - preserving %s pinning to memory %q", c.PrettyName, c.GetCpusetMems())
+			preserveMems, err := parseIDSet(c.GetCpusetMems())
+			if err != nil {
+				log.Error("failed to parse CpusetMems: %v", err)
+			} else {
+				zone := p.allocMem(c, preserveMems, true)
+				log.Debug("  - allocated preserved memory %s", c.PrettyName, zone)
+				c.SetCpusetMems(zone.MemsetString())
```

Comment:
When preservation is required by outside this policy, I think we should never touch CpusetMems, even if it would screw up our bookkeeping somehow. Instead of SetCpusetMems it would be good to log.Error if zone is different from original pinning?

Reply:
That is slightly different and called in libmem a `reservation` as opposed to a container/allocation with `Preserved` inertia/memory. libmem never moves reservations around. Preservations it can move around as a last resort, but libmem guarantees that if a Preserved allocation is moved around, the new memory set is always a superset of the old one. IOW, a preserved allocation never loses any memory node it once had access to.